OSDOCS-19540: adds gRPC support RHCL (#111522)
openshift/openshift-docs: modules/con-about-grpc-support.adoc

// Module included in the following assemblies:
//
// *deployment/deployment.adoc

:_mod-docs-content-type: CONCEPT
[id="con-about-grpc-support_{context}"]
= About GRPCRoute policy attachment

:FeatureName: `GRPCRoute` policy attachment

> include::snippets/technology-preview.adoc[]

Reviewer: I think I more commonly see the TP snippet after the abstract. I'm not sure if it matters, though.

[role="_abstract"]

> Beginning with {prodname} 1.4.0, `GRPCRoute` policy attachment is supported. You can now use `GRPCRoute` custom resources (CRs) for your Remote Procedure Call (gRPC)-native applications and services in all of your {prodname} policies except `TokenRateLimitPolicy` CRs. You can migrate existing policies after you create your `GRPCRoute` CRs.

Reviewer 2: These policies mentioned here don't include extension policies (PlanPolicy, TelemetryPolicy, OIDCPolicy), right? GRPCRoute targeting shouldn't work with all the extension policies we currently have

Author: Correct, because we don't have SDK as Tech Preview yet, so I left those out.

> {prodname} policies can now target `GRPCRoute` resources with the `targetRef` parameter. Specifically, `AuthPolicy` and `RateLimitPolicy` accept `GRPCRoute` as a valid target `kind`. The following well-known attributes are also available for Common Expression Language (CEL) expressions:

Reviewer: Possible to revise this so that `kind` is followed by a noun? https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#object-references

If not no worries.

Author: well, hmmm, `kind` is the key and `GRPCRoute` is the value, so IDK what there is to put after `kind` that won't be wrong (it's not an object, the route is the object)

* `request.grpc.service`: Represents the exact service being called.
* `request.grpc.method`: Represents the exact function being executed.

Using a `GRPCRoute` CR instead of an `HTTPRoute` CR when you already have a gRPC service can be an advantage in the following scenarios:

* Native `service` or `method` matching in routing rules instead of raw path patterns.
* Clearer intent. The route type declares that this is a gRPC service, making the topology more readable.
* Gateway API validation. `GRPCMethodMatch` enforces valid `service` or `method` structure rather than free-form path strings.

The following CRs support `GRPCRoute` CRs in their `targetRef` fields:

* AuthPolicy CRs

> * RateLimitPolicy CRs

Reviewer 2: Same question here. With extensions, this list should at least also include the PlanPolicy

Author: We don't support it yet downstream because of dependency on extension SDK.

// Module included in the following assemblies:
//
// *deployment/deployment.adoc

:_mod-docs-content-type: CONCEPT
[id="con-about-grpc-support_{context}"]

> = About gRCP framework support

Reviewer 3: ```suggestion
= About gRPC framework support
```

> :FeatureName: The gRCP framework

Reviewer 3: ```suggestion
:FeatureName: The gRPC framework
```

The gCRP framework uses `HTTP/2` as its underlying transport protocol and coordinates how messages are sent, handles client or server code generation, and manages features such as streaming, retries, and deadlines. These functions are built-in to the gRCP framework, which means that you do not need to write items such as timeouts into a REST API.

Using a `GRPCRoute` CR instead of an `HTTPRoute` CR can be an advantage in the following scenarios:

* Most of your traffic is internal or mobile-based, and not browser-based.
* You are working with real-time data or large datasets and need to push continuous updates to a client.
* You have microservice-to-microservice communication at a high rate per second. The gRCP framework is especially useful when combined with rate-limit policies in these situations.
* You need long-lived connections.
* You want API changes to be contract-based, predictable, and visible.
* You want fine-grained security and rate-limiting to specific gRPC services and methods rather than the entire Gateway.

The following CRs support `GRPCRoute` CRs in their `targetRef` fields:

** AuthPolicy CRs
** RateLimitPolicy CRs

> ** OIDCPolicy CRs: Additional configuration might apply.

Reviewer 3: OIDCPolicy was not updated to support GRPCRoutes

** PlanPolicy CRs: Additional configuration might apply.

`DNSPolicy` and `TLSPolicy` policies cover attached `GRPCRoute` CRs automatically.

> Telemetry works with `GRPCRoute` CRs with no updates required. While telemetry objects target the `Gateway` object, telemetry captures and reports on traffic flowing through attached `GRPCRoute` CRs. You have the same level of visibility for gRPC traffic as you currently do for HTTP and HTTPS traffic.

Reviewer 3: This doc seems to talk a lot about gRPC itself rather the support we are adding "GRPCRoute policy attachment support"

It comes across as a bit misleading to say "the gRPC framework is supported" and "gRPC framework support" , as if Kuadrant/RHCL is providing gRPC capability. 

What we are actually adding is narrower: **Kuadrant policies can now target GRPCRoute resources via `targetRef`**. Specifically, AuthPolicy, RateLimitPolicy, and PlanPolicy accept `GRPCRoute` as a valid target kind, and new `request.grpc.service` / `request.grpc.method` well-known attributes are available for CEL expressions.

// Module included in the following assemblies:
//
// *deployment/deployment.adoc

:_mod-docs-content-type: CONCEPT
[id="con-about-grpc-support_{context}"]
= About GRPCRoute policy attachment

:FeatureName: `GRPCRoute` policy attachment

include::snippets/technology-preview.adoc[]

[role="_abstract"]

> Beginning with {prodname} 1.4.0, `GRPCRoute` policy attachment is supported. You can now use `GRPCRoute` custom resources (CRs) for your Remote Procedure Call (gRPC)-native applications and services in all of your {prodname} policies except `TokenRateLimitPolicy` CRs. You can migrate existing policies after you create your `GRPCRoute` CRs.

Reviewer 3: OIDCPolicy is also excluded, maybe better to just list the ones that are?

```suggestion
Beginning with {prodname} 1.4.0, GRPCRoute policy attachment is supported. You can attach AuthPolicy and RateLimitPolicy CRs to GRPCRoute resources, enabling authentication and rate-limiting for your gRPC services. TokenRateLimitPolicy CRs do not support GRPCRoute targets.
```

Reviewer 3: I originally suggested this change thinking there were other policies that were also not included, but if we are not talking about SDK policies (PlanPolicy, OIDCPOlicy etc..) this is probably fine.

Author: Well, we list them starting on line 29; let's leave the intro evergreen, then we can just add policies to that list as the support rolls out, WDYT?

Reviewer 3: Yeah OK, i was suggesting the change as you were explicitly listing polices that do not support it here i.e. TokenRateLimitPolicy, but this list was potentially longer than the list that supported it, TokenRateLimitPolicy, PlanPolicy, OIDCPolicy etc...

I'm OK with it as it is if we are not including SDK policies in any of this.

{prodname} policies can now target `GRPCRoute` resources with the `targetRef` parameter. Specifically, `AuthPolicy` and `RateLimitPolicy` accept `GRPCRoute` as a valid target `kind`. The following well-known attributes are also available for Common Expression Language (CEL) expressions:

* `request.grpc.service`: Represents the exact service being called.
* `request.grpc.method`: Represents the exact function being executed.

Using a `GRPCRoute` CR instead of an `HTTPRoute` CR can be an advantage in the following scenarios:

* Most of your traffic is internal or mobile-based, and not browser-based.
* You are working with real-time data or large datasets and need to push continuous updates to a client.
* You have microservice-to-microservice communication at a high rate per second. `GRPCRoute` policy attachment support is especially useful when combined with rate-limiting policies in these situations.
* You need long-lived connections.
* You want API changes to be contract-based, predictable, and visible.

> * You want fine-grained security and rate-limiting to specific gRPC services and methods, but not the entire `Gateway` object.

Reviewer 3: This is mostly describing the benefits of grpc over http rather than GRCPRoute over HTTPRoute.

> You want fine-grained security and rate-limiting to specific gRPC services and methods, but not the entire `Gateway` object.

 Even that is slightly misleading — you can achieve the same with HTTPRoute by matching on path patterns like /Service/Method. The real advantage of GRPCRoute over HTTPRoute for an existing gRPC service is:

  - Native service/method matching in routing rules instead of raw path patterns
  - Clearer intent — the route type declares that this is a gRPC service, making the topology more readable
  - Gateway API validation — GRPCMethodMatch enforces valid service/method structure rather than free-form path strings

 The list should describe why a user who already has a gRPC service would choose a GRPCRoute CR over continuing to route it via an HTTPRoute CR.

The following CRs support `GRPCRoute` CRs in their `targetRef` fields:

** AuthPolicy CRs

> ** RateLimitPolicy CRs

Reviewer 3: ```suggestion
* AuthPolicy CRs
* RateLimitPolicy CRs
```

`DNSPolicy` and `TLSPolicy` policies cover attached `GRPCRoute` CRs automatically.

> Telemetry works with `GRPCRoute` CRs with no updates required. While telemetry objects target the `Gateway` object, telemetry captures and reports on traffic flowing through attached `GRPCRoute` CRs. You have the same level of visibility for gRPC traffic as you currently do for HTTP and HTTPS traffic.

Reviewer 3: ```suggestion
Telemetry works with GRPCRoute CRs with no updates required. Although telemetry objects target the Gateway object, telemetry captures and reports on gRPC traffic matched by attached GRPCRoute CRs. You have the same level of observability for gRPC traffic as for traffic routed through HTTPRoute CRs.
```

It currently sounds like gRPC is an alternative to http, but it is actually all http traffic.

Author: backticks on object names are a docs style requirement

Reviewer 3: Sure, it was more about the wording than backticks, i guess my copy paste just removed them.

// Module included in the following assemblies:
//
// *deployment/deployment.adoc

:_mod-docs-content-type: CONCEPT
[id="con-about-grpc-support_{context}"]
= About GRPCRoute policy attachment

:FeatureName: `GRPCRoute` policy attachment

include::snippets/technology-preview.adoc[]

[role="_abstract"]
Beginning with {prodname} 1.4.0, `GRPCRoute` policy attachment is supported. You can now use `GRPCRoute` custom resources (CRs) for your Remote Procedure Call (gRPC)-native applications and services in all of your {prodname} policies except `TokenRateLimitPolicy` CRs. You can migrate existing policies after you create your `GRPCRoute` CRs.

> {prodname} policies can now target `GRPCRoute` resources with the `targetRef` parameter. Specifically, `AuthPolicy`, `RateLimitPolicy`, and `PlanPolicy` accept `GRPCRoute` as a valid target `kind`. The following well-known attributes are also available for Common Expression Language (CEL) expressions:

Author: ```suggestion
{prodname} policies can now target `GRPCRoute` resources with the `targetRef` parameter. Specifically, `AuthPolicy` and `RateLimitPolicy` accept `GRPCRoute` as a valid target `kind`. The following well-known attributes are also available for Common Expression Language (CEL) expressions:
```
confirm removal (SDK-based)
